### DISCUSSION TITLE
interlaced test output problem with multiroot test runs

### DISCUSSION THREAD

**Discussion by jimasp** [microsoft/vscode-discussions#559] [Extension Development QnA] [answered]
Imagine you have 3 project folders in a multiroot workspace.
You click to run all tests, and all 3 projects begin running their tests.
Each project creates its own `vscode.TestRun` object and will output the stdout from a testrunner using `run.appendOutput`.
It's reasonable to assume that each individual run object would get its own separate run output window, but this is not the case. 
All runs will use the SAME output window.
Thus the output from the 3 project runs is interlaced, and it becomes a garbled mess.

For example, if we have the following code inside a test run handler:
```
const runs: vscode.TestRun[] = [];
const promises: Promise<void>[] = [];
const projectFolders = ["projectA", "projectB", "projectC"];

let i = 0;
for (const project of projectFolders) {
  const runName = project + " run";
  const run = ctrl.createTestRun(request, runName, false);
  runs.push(run);
  promises.push(startTestRunForProject(run, i++ * 3)); // each project runs async
}
await Promise.all(promises);

for (const run of runs) {
  run.end();
}

async function startTestRunForProject(run: vscode.TestRun, pause: number) {
  // run project tests sequentially
  for (let i = 0; i < 6; i++) {
    await new Promise(resolve => setTimeout(resolve, pause)); // (run a test)      
    run.appendOutput(run.name + ": test" + i + " complete\r\n");
  }
}

```
then this produces a SINGLE test terminal window containing interlaced output from all 3 run objects:
```
projectA run: test0 complete
projectA run: test1 complete
projectB run: test0 complete
projectA run: test2 complete
projectA run: test3 complete
projectC run: test0 complete
projectA run: test4 complete
projectB run: test1 complete
projectA run: test5 complete
projectB run: test2 complete
projectC run: test1 complete
projectB run: test3 complete
projectC run: test2 complete
projectB run: test4 complete
projectB run: test5 complete
projectC run: test3 complete
projectC run: test4 complete
projectC run: test5 complete
```

In reality of course test output is more complex, and in the case of BDD, is supposed to be easily readable, which it cannot be if the test output is interlaced.

**Is there a way to separate out the test output from each run, i.e. give each run its own terminal window?**
Or is the only solution to create extension output windows for each project? (which doesn't seem to fit with the design of the vscode test api for test output).  
I don't want to buffer the output as it would give a poor user experience as the user would have to wait until the end of each run to see the output for the run.
Is there a plan to address this, or am I just doing something wrong? How is the user expected to view BDD test output when multiroot projects are running in parallel but limited to using the same terminal window for test output?

Edited to add: 
I've also noticed that if you have two test extensions installed and run all tests, then both extensions use the same terminal window, so it seems? like the test output is tied to the `TestRunRequest` and not the `TestRun`. If true, this is odd conceptually because I am using `run.appendOutput` and not `request.appendOutput`.

Separately, I'm not really clear on the intended relationship between a `vscode.TestRunRequest` and `vscode.TestRun`. The docs simply say "The runHandler should call controller.createTestRun at least once, passing through the original request."  which suggests its up to the extension author.  I'm not sure if creating one run per workspace is a good/supported design or if I should just use one run object for all multiroot projects, i.e. if either approach is better supported/futureproof vs the vscode api. The extension would work either way, I only tried creating separate runs per project to try to address the output problem.

**Comment by gjsjohnmurray** [comment]
@connor4312 might be able to advise.

**Comment by isidorn** [comment]
@jimasp is this still a question that you have? Or have you found a solution?

**Comment by jimasp** [comment, marked as answer]
@isidorn this was fixed by https://github.com/microsoft/vscode/pull/180746

closing
